ArduCopter/surface_tracking.cpp

```diff
@@ -1,67 +1,47 @@
-    }
-    last_update_ms = now;
+    if (((surface == Surface::GROUND) && copter.rangefinder_alt_ok() && (copter.rangefinder_state.glitch_count == 0)) ||
+        ((surface == Surface::CEILING) && copter.rangefinder_up_ok() && (copter.rangefinder_up_state.glitch_count == 0))) {
 
-    // adjust rangefinder target alt if motors have not hit their limits
-    if ((target_rate<0 && !copter.motors->limit.throttle_lower) || (target_rate>0 && !copter.motors->limit.throttle_upper)) {
-        target_dist_cm += dir * target_rate * copter.G_Dt;
-    }
-    valid_for_logging = true;
+        // init based on tracking direction/state
+        RangeFinderState &rf_state = (surface == Surface::GROUND) ? copter.rangefinder_state : copter.rangefinder_up_state;
+        const float dir = (surface == Surface::GROUND) ? 1.0f : -1.0f;
+        offset_cm = copter.inertial_nav.get_altitude() - dir * rf_state.alt_cm;
 
-#if AC_AVOID_ENABLED == ENABLED
-    // upward facing terrain following never gets closer than avoidance margin
-    if (surface == Surface::CEILING) {
-        const float margin_cm = copter.avoid.enabled() ? copter.avoid.get_margin() * 100.0f : 0.0f;
-        target_dist_cm = MAX(target_dist_cm, margin_cm);
+        // reset target altitude if this controller has just been engaged
+        // target has been changed between upwards vs downwards
+        // or glitch has cleared
+        const uint32_t now = millis();
+        if ((now - last_update_ms > SURFACE_TRACKING_TIMEOUT_MS) ||
+            reset_target ||
+            (last_glitch_cleared_ms != rf_state.glitch_cleared_ms)) {
+            copter.pos_control->set_pos_offset_z_cm(offset_cm);
+            reset_target = false;
+            last_glitch_cleared_ms = rf_state.glitch_cleared_ms;
+        }
+        last_update_ms = now;
+        valid_for_logging = true;
+    } else {
+        copter.pos_control->set_pos_offset_z_cm(offset_cm);
     }
-#endif
-
-    // calc desired velocity correction from target rangefinder alt vs actual rangefinder alt (remove the error already passed to Altitude controller to avoid oscillations)
-    const float distance_error = (target_dist_cm - rf_state.alt_cm) - (dir * current_alt_error);
-    float velocity_correction = dir * distance_error * copter.g.rangefinder_gain;
-    velocity_correction = constrain_float(velocity_correction, -SURFACE_TRACKING_VELZ_MAX, SURFACE_TRACKING_VELZ_MAX);
-
-    // return combined pilot climb rate + rate to correct rangefinder alt error
-    return (target_rate + velocity_correction);
 #else
-    return target_rate;
+    copter.pos_control->set_pos_offset_z_cm(offset_cm);
 #endif
+    copter.pos_control->set_pos_target_z_from_climb_rate_cm(target_rate, offset_cm);
 }
 
+
 // get target altitude (in cm) above ground
 // returns true if there is a valid target
-bool Copter::SurfaceTracking::get_target_alt_cm(float &_target_alt_cm) const
+bool Copter::SurfaceTracking::get_target_alt_cm(float &target_dist) const
```

Comment:
get_target_alt_cm()'s arguments are named differently in the .h vs .cpp.  We should keep these consistent.